### PR TITLE
[Serve] Inherit previous deployment num_replicas

### PR DESCRIPTION
**Pull request by simon-mo**
Signed-off-by: simon-mo <simon.mo@hey.com>

## Why are these changes needed?
In Serve autoscaling mode, the new versions of deployment will be resetted to min_replicas instead of the the current number of running replicas. This causes a drop in QPS. 

This PR fixes it and adds a test.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/controller.py
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:371]: Can we change this to

```python
deployment_config.num_replicas = min(
    max(
        previous_deployment.deployment_config.num_replicas,
        autoscaling_config.min_replicas
    ), 
    autoscaling_config.max_replicas
)
```

This would maintain the invariant that `num_replicas` is between `min_replicas` and `max_replicas`.

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_autoscaling_policy.py:925]:
```suggestion
            filters=[("class_name", "=", "ServeReplica:f"), ("state", "=", "DEAD")]
```

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_autoscaling_policy.py:931]: What is this testing for? Is it checking that the second `serve.run` call kills the original 2 deployments?

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (architkulkarni) [inline python/ray/serve/tests/test_autoscaling_policy.py:925]:
```suggestion
            filters=[("class_name", "=", "ServeReplica:f"), ("state", "=", "DEAD")]
```
Perhaps?

**Review by architkulkarni — APPROVED** [review]
Looks good to me, but there might be a typo in the test

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work so far! Could we also add a test to make sure no replicas get killed when we update the autoscaling config to have a higher `max_replicas`.

E.g. if `min_replicas=1`, `max_replicas=10`, and the current `num_replicas=5`, then updating `max_replicas` to 8 shouldn't cause any current replicas to stop after this change is merged in. Updating the `autoscaling_config` is considered a [lightweight change](https://docs.ray.io/en/releases-2.0.1/serve/production-guide/rest-api.html#lightweight-config-updates).

File: python/ray/serve/controller.py
Comment (simon-mo) [reply python/ray/serve/controller.py:371]: I'm pretty sure autoscaling loop will ensure this invariant, so we don't need to do anything to interfere with it. 

https://github.com/ray-project/ray/blob/98da294f0b8fbf0ab9d94ebd552c4d78b4d3aaf8/python/ray/serve/_private/autoscaling_policy.py#L47-L49

File: python/ray/serve/controller.py
Comment (sihanwang41) [inline python/ray/serve/controller.py:367]: [Non blocking merge idea] Is it possible we directly make all these logics (L366 - L373) into the `_set_target_state` of DeploymentState class?

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (simon-mo) [reply python/ray/serve/tests/test_autoscaling_policy.py:931]: yeah the second serve.run should kill them because of the "version" update

Comment (simon-mo) [reply python/ray/serve/tests/test_autoscaling_policy.py:931]: changed to static version

**Review by sihanwang41 — APPROVED** [review]

File: python/ray/serve/controller.py
Comment (simon-mo) [reply python/ray/serve/controller.py:367]: no because the deployment state is not aware of historical deployments (or previous version at the spec)

File: python/ray/serve/controller.py
Comment (architkulkarni) [reply python/ray/serve/controller.py:371]: Ah good point, probably good to not do this in multiple places.  Agree we should add a test for this scenario

File: python/ray/serve/controller.py
Comment (sihanwang41) [reply python/ray/serve/controller.py:367]: We have https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/deployment_state.py#L1993, if the deployment exists, you will be in existing deployment to update your config, so you should have all information.

My two cents:

- only have one place to update the num_replicas which is better inside the DeploymentState.
- Your current change is under the same structure as before, no need to change in this pr :)

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (simon-mo) [inline python/ray/serve/tests/test_autoscaling_policy.py:944]: @shrekris-anyscale ptal

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (simon-mo) [reply python/ray/serve/tests/test_autoscaling_policy.py:939]: static version didn't work.. it will get updated no matter what

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_autoscaling_policy.py:947]: I think this comment is missing something after `Prepare the`.

File: python/ray/serve/tests/test_autoscaling_policy.py
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_autoscaling_policy.py:944]: Nice! This looks good.

**Review by shrekris-anyscale — APPROVED** [review]
Nice work, the new test looks good!

**Comment by scv119** [comment]
@simon-mo looks this breaks windows test

**Comment by simon-mo** [comment]
@scv119 i see. i need to fix CI here instead of revert, please stamp and merge https://github.com/ray-project/ray/pull/29832

**Pull request merged**
